### ISSUE TITLE
azure insights metrics list failure

### ISSUE THREAD

**Issue by djberg96** [Azure/azure-xplat-cli#2698]
CLI Version: 0.9.19
OS Type: Mac
Installation via:  mac installer
Mode: ARM

Description:
The "azure insights metrics list" does not seem to work, nor does it seem to acknowledge command line switches, only positional arguments.

Steps to reproduce:
azure insights metrics list "/subscriptions/xxx/resourceGroups/yyy/providers/Microsoft.Compute/virtualMachines/my_vm" 01:00:00

Error stack trace:

```
2016-03-30T16:51:54.028Z:
{ [Error: No registered resource provider found for location 'eastus' and API version '2014-04-01' for type 'virtualMachines'. The supported api-versions are '2015-05-01-preview, 2015-06-15, 2016-03-30'. The supported locations are 'eastus, eastus2, westus, centralus, northcentralus, southcentralus, northeurope, westeurope, eastasia, southeastasia, japaneast, japanwest, australiaeast, australiasoutheast, brazilsouth'.]
  stack: [Getter/Setter],
  code: 'NoRegisteredProviderFound',
  statusCode: 400,
  requestId: 'ae8c987f-18c5-42be-8dc6-64baf91897bc',
  __frame: 
   { name: 'insightsMetricsCommand__executeCmd__3',
     line: 78,
     file: '/usr/local/azure/lib/commands/arm/insights/insights.metrics.js',
     prev: 
      { name: 'insightsMetricsCommand__prepareAndExecute__2',
        line: 62,
        file: '/usr/local/azure/lib/commands/arm/insights/insights.metrics.js',
        prev: [Object],
        calls: 1,
        active: false,
        offset: 13,
        col: 16 },
     calls: 0,
     active: false,
     offset: 4,
     col: 45 },
  rawStack: [Getter] }
Error: No registered resource provider found for location 'eastus' and API version '2014-04-01' for type 'virtualMachines'. The supported api-versions are '2015-05-01-preview, 2015-06-15, 2016-03-30'. The supported locations are 'eastus, eastus2, westus, centralus, northcentralus, southcentralus, northeurope, westeurope, eastasia, southeastasia, japaneast, japanwest, australiaeast, australiasoutheast, brazilsouth'.
  <<< async stack >>>
  at insightsMetricsCommand__executeCmd__3 (/usr/local/azure/lib/commands/arm/insights/insights.metrics.js:82:46)
  at insightsMetricsCommand__prepareAndExecute__2 (/usr/local/azure/lib/commands/arm/insights/insights.metrics.js:75:17)
  at __1 (/usr/local/azure/lib/commands/arm/insights/insights.metrics.js:41:32)
  <<< raw stack >>>
    at Function.ServiceClient._normalizeError (/usr/local/azure/node_modules/azure-common/lib/services/serviceclient.js:815:23)
    at /usr/local/azure/node_modules/azure-common/lib/services/filters/errorhandlingfilter.js:44:29
    at Request._callback (/usr/local/azure/node_modules/azure-common/lib/http/request-pipeline.js:109:14)
    at Request.self.callback (/usr/local/azure/node_modules/azure-common/node_modules/request/request.js:199:22)
    at emitTwo (events.js:87:13)
    at Request.emit (events.js:172:7)
    at Request.<anonymous> (/usr/local/azure/node_modules/azure-common/node_modules/request/request.js:1160:14)
    at emitOne (events.js:82:20)
    at Request.emit (events.js:169:7)
    at IncomingMessage.<anonymous> (/usr/local/azure/node_modules/azure-common/node_modules/request/request.js:1111:12)
```

**Comment by amarzavery** [comment]
@djberg96 - The error says that you need to register the RP, for your current subscription. This needs to happen only for the first time when you are using a command for an RP that is not registered by default.

```
azure provider list // will show the list of providers registered/un-registered for your current subscription
azure provider register "microsoft.insights"
```
After this when you execute the insights command, it should work.

**Comment by djberg96** [comment]
@amarzavery It says I'm already registered:
```
dberger:~ >azure provider list
info:    Executing command provider list
+ Getting ARM registered providers                                             
data:    Namespace                          Registered   
data:    ---------------------------------  -------------
data:    Microsoft.ApiManagement            Registered   
data:    Microsoft.Batch                    Registered   
data:    Microsoft.ClassicCompute           Registered   
data:    Microsoft.ClassicNetwork           Registered   
data:    Microsoft.ClassicStorage           Registered   
data:    Microsoft.Compute                  Registered   
data:    Microsoft.Insights                 Registered 
...
```

**Comment by amarzavery** [comment]
Ohhh...  The error message says: "No registered resource provider found for location 'eastus' and API version '2014-04-01' "
- region and api-version.

`azure provider show Microsoft.Insights` should tell you which api versions are registered in east us.

Let me loop in the insights team

@gucalder - Can you please help this customer?

**Comment by djberg96** [comment]
@amarzavery The output of "azure provider show Microsoft.Insights" does not appear to contain api-version information, only the resource type name and locations, even with verbose output. And, as far as I can tell, I can't explicitly set the api-version myself anyway.

**Comment by amarzavery** [comment]
With --json it shows the detailed info.
Based on your command usage:
```
azure insights metrics list "/subscriptions/xxx/resourceGroups/yyy/providers/Microsoft.Compute/virtualMachines/my_vm" 01:00:00
```
metricsdefinition would be the resource type to look at:
```
azure provider show Microsoft.insights --json
{
  "id": "/subscriptions/************/providers/microsoft.insights",
  "namespace": "microsoft.insights",
  "registrationState": "NotRegistered",
  "resourceTypes": [
    {
      "resourceType": "components",
      "locations": [
        "Central US"
      ],
      "apiVersions": [
        "2015-05-01",
        "2014-12-01-preview",
        "2014-08-01",
        "2014-04-01"
      ]
    },
    {
      "resourceType": "webtests",
      "locations": [
        "Central US"
      ],
      "apiVersions": [
        "2015-05-01",
        "2014-08-01",
        "2014-04-01"
      ]
    },
 . . .
    {
      "resourceType": "metricDefinitions",
      "locations": [
        "West US",
        "East US",
        "North Europe",
        "West Europe",
        "East Asia",
        "Southeast Asia",
        "Japan East",
        "Japan West",
        "North Central US",
        "South Central US",
        "East US 2",
        "Central US",
        "Brazil South"
      ],
      "apiVersions": [
        "2015-07-01"
      ]
    },
```
looks like for metricDefinitions the version is 2015-07-01
@gucalder - Can you confirm?

**Comment by gucalder** [comment]
To set the context better I have some questions/comments:
1. Is this the first time the customer tries that command or was the command running successfully with previous versions the CLI?
2. Is the customer's subscription already registered with our RP? (This has to be done explicitly using some ARM command)
3. If the role being used by the customer is not the subscription admin, does that role have permissions to access our namespace?

**Comment by djberg96** [comment]
@amarzavery I see the same thing, but I did notice that we had to use 2014-04-01 in our azure-armrest gem or it wouldn't work.

@gucalder, to answer your questions:

1) First time.
2) See above. As far as I can tell, it's already registered.
3) I'm a co-admin on the subscription.

**Comment by gucalder** [comment]
Status of the investigation:

The command above is trying to get the metrics from a V2 virtual machine. Metrics for a V2 VM come from storage (i.e. shoebox), the cli currently does not support getting metrics from shoebox because reading from shoebox requires a thick client that is not available in node.js. This same command should work in the PowerShell version though. PowerShell does support retrieving metrics from shoebox.

There is already work in progress to enable these scenarios for cli, but it is not ready yet.

Definitely, the error message is not helpful and we are going to work to improve it.
We are going to improve our documentation in relation to this too.

**Comment by djberg96** [comment]
@gucalder Ok, thanks. A nicer error message would definitely be helpful. I'll go ahead and close this.

**Comment by djberg96** [comment]
For anyone watching, the `azure insights metrics definition list` will also fail because MS is returning a bogus value for the api-version. This is the error I get:

`No registered resource provider found for location 'westus' and API version '2015-07-01' for type 'virtualMachines/metricDefinitions'. The supported api-versions are '2014-04-01'. The supported locations are 'eastus, eastus2, westus, centralus, northcentralus, southcentralus, northeurope, westeurope, eastasia, southeastasia, japaneast, japanwest, australiaeast, australiasoutheast, brazilsouth, southindia, centralindia, westindia, canadacentral, canadaeast, westus2, westcentralus, uksouth, ukwest'.`

We had to hard code the "2014-04-01" api-version string in our own library, but we've also found that it doesn't work for non-public environments.

I've reported this issue to Microsoft.
